#236 Start page with progress and button to resume

nmadruga opened the pull request:
Fix #236 

~~Branched off [PR of issue #176 Show 404 unreachable page as rollback component](292) please merge it before this one.~~ DONE
~~Using Backend with latest schema changes and some more inserted data: [Back-end#291F](https://github.com/openmsupply/application-manager-server/pull/197)~~ DONE

### Description
- Display progress on application (for the Applicant) when loading the page `application/<serial>` is loaded and the application is still in DRAFT status.
- If application is SUBMITTED will load the `application/<serial>/summary` page
- Otherwise loads the application start page with progress as:
  - 100% progress in a section means - all `required` have been filled out.
  - When all sections have 100% completed fields (and valid) show the **Review & Summary** button
  - **No progress bar** displayed when no fields are filled in section
  - **Green** progress bar if all filled field have are valid
  - **Red** if at least one field is invalid
  - First section with incomplete/invalid fields show the **Resume** button -> that links a specific page

[UPDATED on 05/02/2021]
### Unrelated changes
- Add helper `getAssignedElements` to simplify `useGetReviewAssignment` hook, it retrieves sections and questions assigned to the reviewer
- Add helper `updateSectionsProgress` to update received by props array of **SectionDetails** with the progress in each section after receiving a promise back from the `revalidateAll` method.

### Existing todos:
- ~~Not sure if should display a Start page when there is no `startMessage` set for the application template. For now is going straight to section 1 page 1. But can change this to go to first incomplete/invalid section. I wasn't sure if we would display a start page to the Applicant anyway (not on start, but when continuing the application).~~ Edge case: Goes straight to 1st page in 1st section
- The link on the **Resume** button isn't working. It show the URL to the correct page but doesn't redirect. I think it's a problem with the `ApplicationStart` component not being part in the Router...

CarlosNZ commented:
>     * Not sure if should display a Start page when there is no `startMessage` set for the application template. For now is going straight to section 1 page 1.

This is by design. This way things like User Registration can be launched in to immediately without unnecessary start pages. (I think we discussed this when we discussed the Start message)

CarlosNZ commented on src/utils/helpers/application/revalidateAll.ts:
```diff
@@ -7,23 +7,36 @@ import {
-  elementsState: ApplicationElementStates,
-  responsesByCode: ResponsesByCode,
-  currentUser: User,
+interface RevalidateAllProps {
+  elementsState: ApplicationElementStates
+  responsesByCode: ResponsesByCode
+  currentUser: User
+  sectionCode?: string
+  strict?: boolean
+}
```
Changes here look good and useful. Just checking if you've thoroughly regression tested this method in its other places (Summary page access and Submit), so we don't get any unintended regressions slipping through. (Esp related to the removal of the `strict` flag)

nmadruga replied:
I have tested. It seemed ok, but I think it would be nice if you could check too.

I have reverted the change, so the `strict` check is back

CarlosNZ commented on src/utils/hooks/useGetSectionProgress.tsx:
```diff
@@ -0,0 +1,86 @@
+
+  async function runScanSectionProgress() {
+    let sectionsProgress: SectionsProgress = {}
+    templateSections.forEach(async ({ code, title, index }) => {
```
As discussed, you probably want a `for...in` loop here as it's in an `async` function

CarlosNZ commented on src/utils/hooks/useGetSectionProgress.tsx:
```diff
@@ -0,0 +1,86 @@
+        responsesByCode,
+        currentUser,
+        sectionCode: code,
+        strict: false, // Do we still need this flag?
```
Yes?

CarlosNZ commented on src/utils/hooks/useGetSectionProgress.tsx:
```diff
@@ -0,0 +1,86 @@
+    })
+  }
+
+  const getLinkToSection = (validityFailures: any): string => {
```
Am I correct in my understanding that this will get a link to the Section AND Page? If so, maybe call it `getLinkToLocation` or `getLinkToInProgressLocation` ?

CarlosNZ commented on src/utils/hooks/useLoadApplication.tsx:
```diff
@@ -12,11 +13,13 @@ import {
 
 const useLoadApplication = ({ serialNumber, networkFetch }: UseGetApplicationProps) => {
   const [application, setApplication] = useState<ApplicationDetails>()
+  const [templateType, setTemplateType] = useState<TemplateType>()
```
Maybe call this `templateInfo`? I find `templateType` to imply that it's a Type (as in Typescript), but it isn't.

CarlosNZ commented on src/utils/hooks/useLoadApplication.tsx:
```diff
@@ -68,6 +71,15 @@ const useLoadApplication = ({ serialNumber, networkFetch }: UseGetApplicationPro
 
       setApplication(applicationDetails)
 
+      const { id, code, name, startMessage } = application.template as Template
+
+      setTemplateType({
+        id,
+        code,
+        name: name as string,
+        startMessage: startMessage ? startMessage : undefined,
+      })
+
```
Just wondering why you only want these four fields? Why not just return the whole `application.template` object, even if it's not all used currently?

Also, just a question: where was the PageWrapper getting this info from previously?

nmadruga replied:
This info was not required for PageWrapper, and it's only now to load the `ApplicationStart` component. Before the component was called by `ApplicationCreate` and this information was getting retrieved by `useLoadTemplate`

nmadruga replied:
I don't like so much the idea of returning the Template object because it has so many unusable fields in it, that's why I'm always converting into a local type. I think this can stay the same way

CarlosNZ commented on src/utils/hooks/useLoadApplication.tsx:
```diff
@@ -68,6 +71,15 @@ const useLoadApplication = ({ serialNumber, networkFetch }: UseGetApplicationPro
+        id,
+        code,
+        name: name as string,
+        startMessage: startMessage ? startMessage : undefined,
```
```suggestion
        startMessage,
```
Pretty sure this does the same thing -- if it's undefined the property will carry through the "undefined" value, won't it?

nmadruga replied:
It can be `string` | `undefined` | `Null` => I'm just updating to be either `string` or `undefined`

CarlosNZ commented on src/utils/hooks/useGetSectionProgress.tsx:
```diff
@@ -0,0 +1,86 @@
+        },
+      }
+
+      setSections(sectionsProgress)
```
This is within the `for..each` loop. As well as the async problem mentioned above, wouldn't it be better to collect all the section loop results together and then only do one `setSections` mutation with the whole thing after the loop?

nmadruga replied:
Done: df2a4cf

CarlosNZ commented on src/containers/Application/ApplicationOverview.tsx:
```diff
@@ -86,11 +86,11 @@ const ApplicationOverview: React.FC = () => {
-      currentUser as User
-    )
+    const revalidate = await revalidateAll({
+      elementsState: elementsState as ApplicationElementStates,
+      responsesByCode: responsesByCode as ResponsesByCode,
+      currentUser: currentUser as User,
```
I see -- it's because Typescript won't allow `elementsStateaas ApplicationElementStates`. Ugh, it's so ugly 😆

CarlosNZ commented on src/components/Application/ApplicationStart.tsx:
```diff
@@ -30,8 +37,27 @@ const ApplicationStart: React.FC<ApplicationStartProps> = (props) => {
   }, [startMessage, currentUser])
 
+  const findFirstImcompleteSection = () => {
+    const firstIncompleteSection = Object.entries(sectionsProgress)
```
Also -- same comment as in the other file: if this is actually linking to a section AND page, perhaps call it `firstIncompleteLocation`?

CarlosNZ commented on src/components/Application/ApplicationStart.tsx:
```diff
@@ -60,23 +86,72 @@ const ApplicationStart: React.FC<ApplicationStartProps> = (props) => {
+                              <Button
+                                color="blue"
+                                as={Link}
+                                to={link} // TODO: Link not working - page doesn't refresh!
```
Hmmm, very strange. Would be good to get to the bottom of this in this PR if possible.

nmadruga replied:
Finally!!! Done: c7bd60a

CarlosNZ reviewed (changes requested):
Cool, pretty big issue! I think there's probably some additional streamlining we can do around validation (progress bar, etc.) but we'll need to talk and plan that for another time.

nmadruga commented:
> This is by design. This way things like User Registration can be launched in to immediately without unnecessary start pages. (I think we discussed this when we discussed the Start message) I don't think it matters about going to the "first invalid page", since this functionality was intended for short applications that should be done in one go (like User Registration).

Cool. I'll leave it as it is.

nmadruga commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -214,8 +230,14 @@ const ApplicationPageWrapper: React.FC = () => {
+  ) : loadStart && sections ? (
+    <ApplicationStart
+      template={templateType as TemplateType}
+      sectionsProgress={sections}
```
As discussed in person - I'll be looking to change that in a moment - will just finish replying to your other review suggestions first

nmadruga replied:
Done: 53c0803

nmadruga commented on src/containers/Application/ApplicationPageWrapper.tsx:
```diff
@@ -192,11 +208,11 @@ const ApplicationPageWrapper: React.FC = () => {
-      currentUser as User
-    )
+    const revalidate = await revalidateAll({
+      elementsState: elementsState as ApplicationElementStates,
+      responsesByCode: responsesByCode as ResponsesByCode,
+      currentUser: currentUser as User,
```
It doesn't let you do that : (

nmadruga commented on src/utils/helpers/application/revalidateAll.ts:
```diff
@@ -7,23 +7,36 @@ import {
       elementsState[key].category === 'QUESTION' &&
       elementsState[key].isVisible === true &&
       // Strict/Loose validation logic:
-      ((strict && elementsState[key].isRequired) || responsesByCode[key]?.isValid !== null)
```
As discussed earlier - I'll put it back and use with `strict` more for this use

nmadruga commented:
@CarlosNZ I have done quite a big refactor here, basically:
- Removed use of `SectionProgress` from components
- Renamed the type `TemplateSectionPayload` to `SectionDetails`
- Updated everywhere that was constructing or using the sections to use the new type (and renamed the variables)
- Renamed type `TemplateType` to `TemplateDetails` and everywhere using it (and renamed the variables)
- Create a new **helper** `updateSectionsProgress` to build an `SectionDetails` array from the received one adding the progress fields to it.
- Improved (on this new helper) the loop to retrieve promises and return the result (using a dispatch state hook) when finished.
- Now the `applicationPageWrapper` always set the progress for all sections and only use it (to display in the `ApplicationStart` component if the page is loaded with `application/<serial>`.

nmadruga commented:
I still haven't figured out the link not working though. I will have a look after on this too.

nmadruga commented:
Ok. Now I think it's all ready for a re-review. Sorry it has sooo many changes. I had to change a lot of files to change the types from TemplateSectionPayload to SectionDetails and since I was doing that I tried to make it tidier.

But you might need to have another look in many files, so no rush.

CarlosNZ reviewed (changes requested):
Heya, code looks good -- tidier and easier to follow, so thanks for doing that.

However, I'm getting a crash when I try to load the start page of an in-progress application.

To reproduce:
- Start an application
- Go back to the homepage
- Go to the application start page either via URL or List.

I get a white screen, with lots of console errors. :(

CarlosNZ commented:
> However, I'm getting a crash when I try to load the start page of an in-progress application.

Possibly to do with back-end schema changes in latest `master`?

nmadruga commented:
Thanks @CarlosNZ I'll have a look now

nmadruga commented:
I have ran the generated file and all seems good to me. Please try again after logout/login to refresh the cached queries/data

CarlosNZ commented:
> I have ran the generated file and all seems good to me. Please try again after logout/login to refresh the cached queries/data

Great, thanks. Working nicely now. Must have been and out-dated graphql.tsx at my end.

CarlosNZ reviewed (approved):
Great, working nicely now and all looks good.

Now that this is more well-established, it'd be good to get some internal docs regarding the application start/load process etc.and what all the different variables mean (and what they're used for).
